text/0517-io-os-reform.md

```diff
@@ -473,12 +635,252 @@ throughout IO, we can go on to explore the modules in detail.
+follow `std::iter`. Along the way, it generalizes the `by_ref` adapter:
+
+```rust
+trait ReadExt: Read {
+    // ... eliding the methods already described above
+
+    // Reify a borrowed reader as owned
+    fn by_ref<'a>(&'a mut self) -> ByRef<'a, Self> { ... }
+
+    // Read everything from `self`, then read from `next`
+    fn chain<R: Reader>(self, next: R) -> Chain<Self, R> { ... }
+
+    // Adapt `self` to yield only the first `limit` bytes
+    fn take(self, limit: u64) -> Take<Self> { ... }
+
+    // Whenever reading from `self`, push the bytes read to `out`
+    fn tee<W: Writer>(self, out: W) -> Tee<Self, W> { ... }
+}
+
+trait WriteExt: Write {
+    // ... eliding the methods already described above
+
+    // Reify a borrowed writer as owned
+    fn by_ref<'a>(&'a mut self) -> ByRef<'a, Self> { ... }
+
+    // Whenever bytes are written to `self`, write them to `other` as well
+    fn broadcast<W: Writer>(self, other: W) -> Broadcast<Self, W> { ... }
+}
+
+// An adaptor converting an `Iterator<u8>` to `Read`.
+pub struct IterReader<T> { ... }
+```
+
+As with `std::iter`, these adapters are object unsafe and hence placed
+in an extension trait with a blanket `impl`.
+
+Note that the same `ByRef` type is used for both `Read` and `Write`
+-- and this RFC proposes to use it for `std::iter` as well. The
+insight is that there is no difference between the *type* used for
+by-ref adapters in any of these cases; what changes is just which
+trait defers through it. So, we propose to add the following to `core::borrow`:
+
+```rust
+pub struct ByRef<'a, Sized? T:'a> {
+    pub inner: &'a mut T
+}
+```
+
+which will allow `impl`s like the following in `core::io`:
+
+```rust
+impl<'a, W: Write> Write for ByRef<'a, W> {
+    #[inline]
+    fn write(&mut self, buf: &[u8]) -> Result<uint, W::Err> { self.inner.write(buf) }
+
+    #[inline]
+    fn flush(&mut self) -> Result<(), W::Err> { self.inner.flush() }
+}
+```
+
+#### Free functions
+[Free functions]: #free-functions
+
+The current `std::io::util` module also includes a number of primitive
+readers and writers, as well as `copy`. These are updated as follows:
+
+```rust
+// A reader that yields no bytes
+fn empty() -> Empty; // in theory just returns `impl Reader`
+
+// A reader that yields `byte` repeatedly (generalizes today's ZeroReader)
+fn repeat(byte: u8) -> Repeat;
+
+// A writer that ignores the bytes written to it (/dev/null)
+fn sink() -> Sink;
+
+// Copies all data from a Reader to a Writer
+pub fn copy<E, R, W>(r: &mut R, w: &mut W) -> Result<(), E> where
+    R: Reader<Err = E>,
+    W: Writer<Err = E>
+```
+
+#### Seeking
+[Seeking]: #seeking
+
+The seeking infrastructure is largely the same as today's, except that
+`tell` is renamed to follow the RFC's design principles and the `seek`
+signature is refactored with more precise types:
+
+```rust
+pub trait Seek {
+    type Err;
+    fn position(&self) -> Result<u64, Err>;
+    fn seek(&mut self, pos: SeekPos) -> Result<(), Err>;
```

Comment:
I believe that `lseek` returns the position within the file after the operation is complete, perhaps this could return `u64` as well